File: 0646-maximum-length-of-pair-chain/0646-maximum-length-of-pair-chain.cpp
class Solution {
public:
    bool static compare(vector<int> &v1, vector<int> &v2)
    {
        if(v1[0]==v2[0])
            return v1[1]<v1[1];
        return v1[0]<v2[0];
    }
    int search(vector<int> &arr, int a)
    {
        int l=0;
        int h=arr.size()-1;
        while(l<=h)
        {
            // cout<<l<<" "<<h<<endl;
            int mid=(l+h)/2;
            if(arr[mid]==a)
                return mid;
            else if(arr[mid]<a)
            {
                l=mid+1;
            }
            else
                h=mid-1;
        }
    
        return l;
    }
    int findLongestChain(vector<vector<int>>& pairs) {
        vector<int>height;
        sort(pairs.begin(),pairs.end(),compare);
        height.push_back(INT_MIN);
        int ma=0;
        for(int i=0;i<pairs.size();i++)
        {
           int a=search(height,pairs[i][0]);
             cout<<a<<" ";
            if(a==height.size())
                height.push_back(pairs[i][1]);
            ma=ma>height.size()?ma:height.size();
            height[a]=min(height[a],pairs[i][1]);
            
        }
        return ma-1;
        
    }
};

File: 0646-maximum-length-of-pair-chain/0646-maximum-length-of-pair-chain.java
class Solution {
    static int search(List<Integer>arr, int target)
    {
        int i=0;
        int j=arr.size()-1;
        while(i<=j)
        {
            int mid=(i+j)/2;
            if(arr.get(mid)==target)
                return mid;
            if(arr.get(mid)>target)
                j=j-1;
            else
                i=i+1;
        }
        return i;
        
    }
     public static void sortbyColumn(int arr[][], int col)
    {
       Arrays.sort(arr, (a, b) -> Integer.compare(a[col],b[col])); // increasing order  
    }
    public int findLongestChain(int[][] pairs) {
        sortbyColumn(pairs,0);
        List<Integer> dp=new ArrayList<Integer>();
        dp.add(pairs[0][1]);
        int ans=1;
        for(int i=1;i<pairs.length;i++)
        {
            int index=search(dp,pairs[i][0]);
           
            
            if(index==dp.size())
                dp.add(pairs[i][1]);
            else
                dp.set(index,Math.min(dp.get(index),pairs[i][1]));
            // System.out.println(index +" "+dp.size());
            ans=Math.max(ans,dp.size());
            
        }
        return ans;
        
        
        
    }
}